### DISCUSSION TITLE
CPU more than 100%

### DISCUSSION THREAD

**Discussion by gsamsonov1989** [async-profiler/async-profiler#822] [Q&A] [answered]
Hello, we run jprofiler for CPU and allocations and collecting results into a JFR file.  After examining jfr file we can see that  application + JVM CPU usage is reported as 120%  but overall machine CPU is reported at 100%.  This seems not ok. I'd like to know how async profiler collects this percentage stats (overall machine CPU and jvm (user + system) CPU).  JVM runs in a docker container

**Comment by franz1981** [comment]
as said elsewhere, the defaul way to collect profiles is via cpu-clock events, per thread, unless jprofiler is making uses of itimer, which seems not given that you have *more* samples in the jfr profiling data than via cpu statistics collected outside of async-profiler...
My guess is that machine CPU usage is not accurate enough (it's an average? it access the same stats from the CPU? how?), but cpu-clock perf events should be decently accurate (although they don't include some interrupt handling, meaning you can still have *fever* cpu usage perception than the real one).

**Reply by gsamsonov1989** [reply]
@franz1981 i am sorry mate i don't understand your answer. Yes we use itimer and it generates profiling signal every 10ms interval. But then you lost me. What is `you have more samples in the jfr profiling data than via cpu statistics collected outside of async-profiler` ?  What is `it's an average? it access the same stats from the CPU? how?`  ?  Question is how exactly  CPU percentage usages are calcualted, for machine and for jvm (user + system)

**Reply by franz1981** [reply]
Afaik async profiler, with itimer, can and will loose samples, see https://github.com/async-profiler/async-profiler/issues/272#issuecomment-1093056181

And, additionally, if kernel stacks cannot be collected, it won't be able to report the kernel CPU consuming ones too, further reducing the amount of samples.

**Reply by gsamsonov1989** [reply]
And the loss of samples somehow leads to CPU usage (application + jvm) being 120% and overall machine usage being 100% ?

**Comment by apangin** [comment]
> we run jprofiler

What does jprofiler have to do with this?!

> After examining jfr file we can see that application + JVM CPU usage is reported as 120%

Can you show an example .jfr?

**Reply by gsamsonov1989** [reply]
@apangin  I am sorry i meant async profiler, not jprofiler.

Unfortunately i can't show becuse of legal reasons. 

We assumed it's because there is only 1 core reported from the JVM.  There existed a bug some time when JVM in docker reported only 1 core available instead of all the cores available on the machine.  We assumed async profiler collects these events from JVM and that's that.

**Reply by apangin** [reply, marked as answer]
Right, async-profiler previously obtained the number of available processors from the JVM.
This has been changed in #751 - now it shows whatever OS reports. So the bug is likely fixed already.
